File: Examples/Getting-Started/Page-Through-Results/Python/page_through_results.py
import json
import os
from typing import Protocol

from netld_example_client import NetLDClient, NetLDError


class ChangeLogPager(Protocol):
    def get_configuration_change_log_page(
        self, network: str, ip_address: str, offset: int, page_size: int
    ) -> dict: ...


def get_all_change_logs(
    client: ChangeLogPager, network: str, ip_address: str, page_size: int
) -> list[dict]:
    if page_size <= 0:
        raise NetLDError("NETLD_PAGE_SIZE must be a positive integer.")

    change_logs: list[dict] = []
    offset = 0
    total: int | None = None

    while total is None or offset < total:
        page = client.get_configuration_change_log_page(
            network, ip_address, offset, page_size
        )
        page_logs = page["changeLogs"]
        page_offset = int(page.get("offset", offset))
        reported_total = int(page.get("total", page_offset + len(page_logs)))
        total = reported_total if total is None else max(total, reported_total)
        change_logs.extend(page_logs)
        next_offset = page_offset + len(page_logs)

        print(
            f"Fetched {len(page_logs)} records at offset {page_offset} "
            f"({next_offset} of {total})"
        )

        if next_offset >= total:
            break
        if not page_logs or next_offset <= offset:
            raise NetLDError("Paging stopped before all results were returned.")
        offset = next_offset

    return change_logs


def main() -> None:
    client = NetLDClient.from_env()
    network = os.environ.get("NETLD_NETWORK", "Default")
    ip_address = os.environ.get("NETLD_DEVICE_IP", "").strip()
    if not ip_address:
        raise NetLDError("Set NETLD_DEVICE_IP in .env before running this example.")
    try:
        page_size = int(os.environ.get("NETLD_PAGE_SIZE", "10"))
    except ValueError as error:
        raise NetLDError("NETLD_PAGE_SIZE must be a positive integer.") from error

    client.login()
    change_logs = get_all_change_logs(client, network, ip_address, page_size)
    print(json.dumps({"total": len(change_logs), "changeLogs": change_logs}, indent=2))


if __name__ == "__main__":
    main()

File: Examples/Getting-Started/Page-Through-Results/nodeJS/page-through-results.mjs
import { NetLDClient, NetLDError } from "./netld-example-client.mjs";
import { pathToFileURL } from "node:url";

export async function getAllChangeLogs(client, network, ipAddress, pageSize) {
  if (!Number.isInteger(pageSize) || pageSize <= 0) {
    throw new NetLDError("NETLD_PAGE_SIZE must be a positive integer.");
  }

  const changeLogs = [];
  let offset = 0;
  let total = null;

  while (total === null || offset < total) {
    const page = await client.getConfigurationChangeLogPage(
      network,
      ipAddress,
      offset,
      pageSize,
    );
    const pageOffset = Number(page.offset ?? offset);
    const reportedTotal = Number(page.total ?? pageOffset + page.changeLogs.length);
    total = total === null ? reportedTotal : Math.max(total, reportedTotal);
    changeLogs.push(...page.changeLogs);
    const nextOffset = pageOffset + page.changeLogs.length;

    console.log(
      `Fetched ${page.changeLogs.length} records at offset ${pageOffset} ` +
        `(${nextOffset} of ${total})`,
    );

    if (nextOffset >= total) break;
    if (page.changeLogs.length === 0 || nextOffset <= offset) {
      throw new NetLDError("Paging stopped before all results were returned.");
    }
    offset = nextOffset;
  }

  return changeLogs;
}

export async function main() {
  const client = NetLDClient.fromEnv();
  const network = process.env.NETLD_NETWORK || "Default";
  const ipAddress = (process.env.NETLD_DEVICE_IP || "").trim();
  if (!ipAddress) {
    throw new NetLDError("Set NETLD_DEVICE_IP in .env before running this example.");
  }
  const pageSize = Number.parseInt(process.env.NETLD_PAGE_SIZE || "10", 10);

  await client.login();
  const changeLogs = await getAllChangeLogs(client, network, ipAddress, pageSize);
  console.log(JSON.stringify({ total: changeLogs.length, changeLogs }, null, 2));
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  await main();
}
